arch/xtensa/core/mpu.c

```diff
@@ -701,3 +707,419 @@ void xtensa_mpu_init(void)
+	 * here and let the application set this instead.
+	 */
+	return CONFIG_MAX_DOMAIN_PARTITIONS;
+}
+
+int arch_mem_domain_partition_remove(struct k_mem_domain *domain,
+				     uint32_t partition_id)
+{
+	int ret;
+	uint32_t perm;
+	bool readable, writable, executable;
+	struct xtensa_mpu_map *map = &domain->arch.mpu_map;
+	struct k_mem_partition *partition = &domain->partitions[partition_id];
+	uintptr_t end_addr = partition->start + partition->size;
+
+	if (end_addr <= partition->start) {
+		ret = -EINVAL;
+		goto out;
+	}
+
+	/*
+	 * This is simply to get rid of the user permissions and retain
+	 * whatever the kernel permissions are. So that we won't be
+	 * setting the memory region permission incorrectly, for example,
+	 * marking read only region writable.
+	 *
+	 * Note that Zephyr does not do RWX partitions so we can treat it
+	 * as invalid.
+	 */
+	decode_kernel_permission(partition->attr, &readable, &writable, &executable);
+
+	if (writable) {
+		if (readable) {
+			perm = XTENSA_MPU_ACCESS_P_RW_U_NA;
+		} else {
+			perm = XTENSA_MPU_ACCESS_P_WO_U_WO;
+		}
+	} else if (executable) {
+		if (readable) {
+			perm = XTENSA_MPU_ACCESS_P_RX_U_NA;
+		} else {
+			perm = XTENSA_MPU_ACCESS_P_X_U_NA;
+		}
+	} else if (readable) {
+		perm = XTENSA_MPU_ACCESS_P_RO_U_NA;
+	} else {
+		/* fail safe to have no access at all. */
+		perm = XTENSA_MPU_ACCESS_P_NA_U_NA;
+	}
+
+	/*
+	 * Reset the memory region attributes by simply "adding"
+	 * a region with default attributes. If entries already
+	 * exist for the region, the corresponding entries will
+	 * be updated with the default attributes. Or new entries
+	 * will be added to carve a hole in existing regions.
+	 */
+	ret = mpu_map_region_add(map, partition->start, end_addr,
+				 perm,
+				 CONFIG_XTENSA_MPU_DEFAULT_MEM_TYPE,
+				 NULL);
+
+out:
+	return ret;
+}
+
+int arch_mem_domain_partition_add(struct k_mem_domain *domain,
+				  uint32_t partition_id)
+{
+	int ret;
+	struct xtensa_mpu_map *map = &domain->arch.mpu_map;
+	struct k_mem_partition *partition = &domain->partitions[partition_id];
+	uintptr_t end_addr = partition->start + partition->size;
+
+	if (end_addr <= partition->start) {
+		ret = -EINVAL;
+		goto out;
+	}
+
+	ret = mpu_map_region_add(map, partition->start, end_addr,
+				 (uint8_t)partition->attr,
+				 CONFIG_XTENSA_MPU_DEFAULT_MEM_TYPE,
+				 NULL);
+
+out:
+	return ret;
+}
+
+int arch_mem_domain_thread_add(struct k_thread *thread)
+{
+	int ret = 0;
+
+	/* New memory domain we are being added to */
+	struct k_mem_domain *domain = thread->mem_domain_info.mem_domain;
+
+	/*
+	 * this is only set for threads that were migrating from some other
+	 * memory domain; new threads this is NULL.
+	 */
+	struct xtensa_mpu_map *old_map = thread->arch.mpu_map;
+
+	bool is_user = (thread->base.user_options & K_USER) != 0;
+	bool is_migration = (old_map != NULL) && is_user;
+
+	uintptr_t stack_end_addr = thread->stack_info.start + thread->stack_info.size;
```

Comment:
Isn't the automatic addition of thread stacks to a memory domain handled by the kernel, not the arch layer?  I swear I think I've seen code to do this, but might be wrong.  If not, it really needs to be.  We really don't want security policy like this being done incompatibly between architectures.

Reply:
This is to figure out if we are at the end of memory, where if we add them together would wrap back to 0. The MPU uses 0xFFFFFFFFU as end of memory and not 0. This is in the next few lines.